distributed/worker.py

```diff
@@ -2562,18 +2563,28 @@ def _transition(
                 recs, instructions = self._transition(
                     ts, "released", stimulus_id=stimulus_id
                 )
-                v = recs.get(ts, (finish, *args))
                 v_state: str
                 v_args: list | tuple
-                if isinstance(v, tuple):
-                    v_state, *v_args = v
-                else:
-                    v_state, v_args = v, ()
-                b_recs, b_instructions = self._transition(
-                    ts, v_state, *v_args, stimulus_id=stimulus_id
+                while v := recs.pop(ts, None):
+                    if isinstance(v, tuple):
+                        v_state, *v_args = v
+                    else:
+                        v_state, v_args = v, ()
+                    if v_state == "forgotten":
+                        # We do not want to forget. The purpose of this
+                        # transition path is to get to `finish`
+                        continue
```

Comment:
This feels sort of weird (if the transition function is telling us to go to `forgotten`, but it shouldn't, isn't that an error in the transition function?). But I think I don't know the worker state machine well enough to judge.

Reply:
Well, this path is a bit of a shortcut to not have to implement all transitions. For instance, if we want to transition `error -> waiting` this is not supposed to be a direct transition (see also https://github.com/dask/distributed/issues/4413#issuecomment-817820890) but it is supposed to transition over released.
By not implementing a `error->waiting` transition, we'll force the engine to "reset" the task and transition to released first, i.e.
`error->warning`
will break out into
`error -> released`
`released -> waiting`

The transition function itself cannot (or should not?) have this kind of global information